File: androidApp/src/main/java/com/leoleo/kmmgithubsearch/android/ui/detail/DetailViewModel.kt
package com.leoleo.kmmgithubsearch.android.ui.detail

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.leoleo.kmmgithubsearch.di.Container
import com.leoleo.kmmgithubsearch.domain.repository.GithubRepoRepository
import kotlinx.coroutines.CoroutineExceptionHandler
import kotlinx.coroutines.launch

class DetailViewModel constructor(
    private val repository: GithubRepoRepository = Container.githubRepository
) : ViewModel() {

    var uiState by mutableStateOf<UiState>(UiState.Initial)
        private set
    private val coroutineExceptionHandler = CoroutineExceptionHandler { _, throwable ->
        uiState = UiState.Error(throwable)
    }

    fun getRepositoryDetail(ownerName: String, repositoryName: String) {
        viewModelScope.launch(coroutineExceptionHandler) {
            uiState = UiState.Loading
            uiState = UiState.Data(repository.getRepositoryDetail(ownerName, repositoryName))
        }
    }
}

File: iosApp/iosApp/UI/Detail/DetailViewModel.swift
//
//  DetailViewModel.swift
//  iosApp
//
//  Created by LeoAndo on 2023/03/14.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import shared

@MainActor
final class DetailViewModel : ObservableObject {
    @Published var uiState: DetailUiState = DetailUiState.initial
    private let repository = Container().githubRepository
    
    func getRepositoryDetail(name: String, ownerName: String) {
        uiState = DetailUiState.loading
        repository.getRepositoryDetail(ownerName: ownerName, repositoryName: name) { result, error in
            if let error = error {
                self.uiState = DetailUiState.error(error.localizedDescription)
            }
            if let result = result {
                self.uiState = DetailUiState.data(DetailUiState.Data(detail:result))
            }
        }
    }
}
